rdrf/rdrf/admin.py

```diff
@@ -129,93 +161,133 @@ def export_registry(registry, request):
+        if unchanged:
+            messages.info(request, _(f"'{', '.join(desc for desc in unchanged)}' already enabled registration"))
 
-    def get_queryset(self, request):
-        if not request.user.is_superuser:
-            user = get_user_model().objects.get(username=request.user)
-            return Registry.objects.filter(registry__in=[reg.id for reg in user.registry.all()])
+    enable_registration_action.short_description = _("Enable registration")
 
-        return Registry.objects.all()
+    def disable_registration_action(self, request, registry_models_selected):
+        unchanged = []
 
-    def has_add_permission(self, request, *args, **kwargs):
-        if request.user.is_superuser:
-            return True
-        return False
+        for registry in registry_models_selected:
+            if not registry.has_feature(RegistryFeatures.REGISTRATION):
+                unchanged.append(str(registry))
+            else:
+                registry.remove_feature(RegistryFeatures.REGISTRATION)
+                registry.save()
 
-    def has_change_permission(self, request, *args, **kwargs):
-        if request.user.is_superuser:
-            return True
-        return False
+                messages.success(request, _(f"Registration disabled for {registry}"))
 
-    def has_delete_permission(self, request, *args, **kwargs):
-        if request.user.is_superuser:
-            return True
-        return False
+        if unchanged:
+            messages.info(request, _(f"'{', '.join(desc for desc in unchanged)}' already had registration disabled"))
 
-    def get_urls(self):
-        original_urls = super(RegistryAdmin, self).get_urls()
+    disable_registration_action.short_description = _("Disable registration")
 
-        return original_urls
+    def create_notifications_action(self, request, registry_models_selected):
+        for registry in registry_models_selected:
+            # TODO: Add default notifications for all registration types. Needs different recipients.
+            try:
+                with transaction.atomic():
+                    notification = EmailNotification(
+                        description=EventType.NEW_PATIENT,
+                        registry=registry,
+                        recipient="{{ patient.user.email }}",
+                    )
+                    notification.save()
 
-    def get_readonly_fields(self, request, obj=None):
-        "Registry code is readonly after creation"
-        return () if obj is None else ("code",)
+                    notification.email_templates.set(
+                        EmailTemplate.objects.filter(default_for_notification=EventType.NEW_PATIENT)
+                    )
+                    notification.save()
+
+                    messages.success(request, render_to_string("admin/notifications_added.html", {
+                        "registry": registry,
+                    }))
+            except IntegrityError as e:
```

Comment:
What special case are we trying to handle here?
Why are we catching `IntegrityError`s and only `IntegrityError`s?

Reply:
I believe this is a holdover from when it was creating notifications and templates in a single action, but I think it may still be useful here.

As we haven't configured requests to be atomic transactions, I've occasionally run into issues during development where some objects are created during a request, but a failure occurs before the request can complete successfully, and these incomplete objects are left hanging before their relationships to other newly-created objects can be specified.

In cases like this one where we're `save()`ing twice, I think it's a good idea to guarantee that if something goes wrong between saves, things will be rolled back properly.

The `IntegrityError` is intended to catch issues with foreign keys, but we could make it more generic and use `DatabaseError`. Or we can remove the transaction altogether if you'd prefer.